src/server.toit

```diff
@@ -106,26 +106,33 @@ class Server:
     response_writer.write_headers STATUS_SWITCHING_PROTOCOLS
     return WebSocket response_writer.detach --no-client
 
+  // Returns true if the connection was detached, false if it was closed.
   run_connection_ connection/Connection handler/Lambda logger/log.Logger -> bool:
     if call_in_finalizer_: connection.call_in_finalizer_ = call_in_finalizer_
     while true:
       request := null
       with_timeout read_timeout:
         request = connection.read_request
-      if not request: return false
+      if not request: return false  // Timed out waiting for request from client.
       request_logger := logger.with_tag "path" request.path
       request_logger.debug "incoming request"
       writer ::= ResponseWriter connection request request_logger
-      try:
-        handler.call request writer
-      finally: | is_exception exception |
-        // Drain unread content to allow the connection to be reused.
-        if writer.detached_: return true
-        if is_exception:
-          writer.close_on_exception_ "Internal Server error - $exception.value.stringify"
-        else:
-          request.drain
-          writer.close
+      unwind_block := : | exception |
+        // If there's an error we can either send a 500 error message or close
+        // the connection.  This depends on whether we had already sent the
+        // headers - can't send a 500 if we already sent a success header.
+        closed := writer.close_on_exception_ "Internal Server error - $exception"
+        closed   // Unwind if the connection is dead.
+      error := catch --trace --unwind=unwind_block:
```

Comment:
Do you always want to get a stack trace? That seems annoying. If you unwind the caller can still produce one -- and if you don't then maybe you shouldn't make noise?

Reply:
This is the point where the library's code gets called.  I think it's OK to always trace, since this will make debugging much easier for the users.  If they don't like it they can wrap their whole router in a catch --unwind=true --trace=false.

Reply:
Does that work? `catch --unwind=true --trace=false`